% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/Freq.R
\name{featureFreq}
\alias{featureFreq}
\title{Extraction of the \emph{K}-Mer Features of RNA and Protein Sequences}
\usage{
featureFreq(seqRNA, seqPro, label = NULL,
  featureMode = c("concatenate", "combine"), computePro = c("RPISeq",
  "DeNovo", "rpiCOOL"), k.Pro = 3, k.RNA = 4, normalize = c("none",
  "row", "column"), normData = NULL, parallel.cores = 2)
}
\arguments{
\item{seqRNA}{RNA sequences loaded by function \code{\link[seqinr]{read.fasta}} of package
"seqinr" (\code{\link[seqinr]{seqinr-package}}). Or a list of RNA sequences.
RNA sequences will be converted into lower case letters.
Each sequence should be a vector of single characters.}

\item{seqPro}{protein sequences loaded by function \code{\link[seqinr]{read.fasta}} of package
"seqinr" (\code{\link[seqinr]{seqinr-package}}). Or a list of protein sequences.
Protein sequences will be converted into upper case letters.
Each sequence should be a vector of single characters.}

\item{label}{optional. A string that indicates the class of the samples such as
"Interact", "Non.Interact". Default: \code{NULL}}

\item{featureMode}{a string that can be \code{"concatenate"} or \code{"combine"}.
If \code{"concatenate"}, the \emph{k}-mer features of RNA and proteins will be simply concatenated.
If \code{"combine"}, the returned dataset will be formed by combining the \emph{k}-mer features of RNA and proteins.
See details below. Default: \code{"concatenate"}.}

\item{computePro}{a string that specifies the computation mode of protein sequence: \code{"RPISeq"}, \code{"DeNovo"},
or \code{"rpiCOOL"}. Ignored when \code{seqType = "RNA"}.
Three modes indicate three different amino acid residues classifications that corresponds to the methods "RPISeq", "De Novo prediction",
and "rpiCOOL". See details below. Default: \code{"RPISeq"}.}

\item{k.Pro}{an integer that indicates the sliding window step of RNA sequences. Default: \code{4}.}

\item{k.RNA}{an integer that indicates the sliding window step of protein sequences. Default: \code{3}.}

\item{normalize}{can be \code{"none"}, \code{"row"} or \code{"column"}. Whether the frequencies should be normalized.
If normalize, the features should be normalized based on each row (sequence) or each column (feature)?
See details below. Default: \code{"none"}.}

\item{normData}{is the normalization data generated by this function.
Only used to extract features of test set and \code{normalize = "column"}.
If the input dataset is training set, or normalize strategy is \code{"none"} or \code{"row"},
just leave \code{normData = NULL}. The normalization data will be computed
based on the input dataset.}

\item{parallel.cores}{an integer specfying the number of cores for parallel computation. Default: \code{2}.
Set \code{parallel.cores = -1} to run with all the cores.}
}
\value{
This function returns a data frame. Row names are the sequences names, and column names are the polymer names.
The names of RNA and protein sequences are seperated with ".",
i.e. the row names format: "\emph{RNASequenceName}.\emph{proteinSequenceName}" (e.g. "YDL227C.YOR198C").
If \code{featureMode = "combine"}, the polymers of RNA and protein sequences are also seperated with ".",
i.e. the column format: "\emph{RNAPolymerName}.\emph{proteinPolymerName}" (e.g. aa.CCA).
}
\description{
Basically a wrapper for \code{\link{computeFreq}} function.
This function can calculate the \emph{k}-mer frequencies of RNA and protein sequences at the same time
and format the results as the dataset that can be used to build classifier.
}
\details{
see \code{\link{computeFreq}}.
}
\section{References}{

[1] Han S, Liang Y, Li Y, \emph{et al}.
ncProR: an integrated R package for effective ncRNA-protein interaction prediction.
(\emph{Submitted})

[2] Shen J, Zhang J, Luo X, \emph{et al}.
Predicting protein-protein interactions based only on sequences information.
Proc. Natl. Acad. Sci. U. S. A. 2007; 104:4337-41

[3] Muppirala UK, Honavar VG, Dobbs D.
Predicting RNA-protein interactions using only sequence information.
BMC Bioinformatics 2011; 12:489

[4] Wang Y, Chen X, Liu Z-P, \emph{et al}.
De novo prediction of RNA-protein interactions from sequence information.
Mol. BioSyst. 2013; 9:133-142

[5] Akbaripour-Elahabad M, Zahiri J, Rafeh R, \emph{et al}.
rpiCOOL: A tool for In Silico RNA-protein interaction detection using random forest.
J. Theor. Biol. 2016; 402:1-8
}

\examples{
data(demoPositiveSeq)
seqsRNA <- demoPositiveSeq$RNA.positive
seqsPro <- demoPositiveSeq$Pro.positive

dataset1 <- featureFreq(seqRNA = seqsRNA, seqPro = seqsPro, label = "Interact",
                        featureMode = "comb", computePro = "DeNovo", k.Pro = 3,
                        k.RNA = 2, normalize = "row", parallel.cores = 2)

dataset2 <- featureFreq(seqRNA = seqsRNA, seqPro = seqsPro, featureMode = "conc",
                        computePro = "rpiCOOL", k.Pro = 3, k.RNA = 4,
                        normalize = "column", parallel.cores = 2)

dataset3 <- featureFreq(seqRNA = seqsRNA, seqPro = seqsPro, featureMode = "conc",
                        computePro = "rpiCOOL", k.Pro = 3, k.RNA = 4,
                        normalize = "column", normData = dataset2$normData,
                        parallel.cores = 2)

}
\seealso{
\code{\link{computeFreq}}
}
